packages/latte-phpstan-compiler/src/PhpParser/NodeVisitor/LinkNodeVisitor.php

```diff
@@ -90,12 +130,28 @@ private function isMethodCallUiControlLink(MethodCall $methodCall): bool
         }
 
         $propertyFetchName = $this->simpleNameResolver->getName($propertyFetch->name);
-        if ($propertyFetchName !== 'uiControl') {
+        if (! in_array($propertyFetchName, ['uiControl', 'uiPresenter'], true)) {
             return false;
         }
         return true;
     }
 
+    /**
+     * @return string[]
+     */
+    private function createMethodNames(string $presenterClassName, string $actionName): array
+    {
+        $targetMethodNames = [];
+        // both methods have to have same parameters, so we check them both if exist
+        foreach (['action', 'render'] as $type) {
+            $targetMethodName = $type . ucfirst($actionName);
+            if (method_exists($presenterClassName, $targetMethodName)) {
+                $targetMethodNames[] = $targetMethodName;
+            }
+        }
+        return $targetMethodNames;
+    }
```

Comment:
Could you split this onto 2 node visitor to lower the complexity? 
It would be easier to debug and test just one of those links.

Reply:
Hm I can, but this is the only change:
```
        if (str_ends_with($targetName, '!')) {
            $targetMethodNames[] = 'handle' . ucfirst(substr($targetName, 0, -1));
            $methodCallVar = new Variable('actualClass');
        } elseif (str_contains($targetName, ':')) {
            $actionParts = explode(':', $targetName);
            $actionName = array_pop($actionParts);
            $presenterName = implode($actionParts);
            $presenterClassName = $this->presenterFactory->formatPresenterClass($presenterName);

            $presenterNameVariable = lcfirst($presenterName) . 'Presenter';
            $methodCallVar = new Variable($presenterNameVariable);

            $attributes['comments'][] = new Doc(
                '/** @var ' . $presenterClassName . ' $' . $presenterNameVariable . ' */'
            );
            $targetMethodNames = $this->createMethodNames($presenterClassName, $actionName);
        }
```

Rest is the same for both.

Reply:
It would be better, as well need to separate between action and render calls. We have a code with same action/render name but different parameters.

Reply:
> That looks pretty complex at first sight

It is not, it is just work with strings


> It would be better, as well need to separate between action and render calls priority. We have some code with same action/render name but different parameters.

Hmmm this would ruin whole concept :D I'm using them (actionX and renderX) always with the same parameters. I can't imagine how would I decide which method to call when parameters are different

Reply:
can you please send me some example of your usage? I'm trying to use it, but I got several errors. Here is my code:

```
public function actionDefault(string $xxx): void
{
}

public function renderDefault(string $yyy): void
{
}
```

So in all requests I have to use `?xxx=aaa&yyy=bbb`. Thats OK and it works. But how to generate link to default in latte? 

With:
```
{link Homepage:default, 'aaa'}
```
It generates Link ?xxx=aaa, so after click I got error `Missing parameter $yyy required by App\Presenters\HomepagePresenter::renderDefault()`

With:
```
{link Homepage:default, 'xxx', 'yyy'}
```
I got: `Invalid link: Passed more parameters than method App\Presenters\HomepagePresenter::actionDefault()`

Reply:
If we solve it after this PR, it would be better.

Reply:
ok, so can we merge this now and start discussion in separate issue / PR?

Reply:
The first comment needs to be resolved first
https://github.com/symplify/symplify/pull/3810#discussion_r762715331

Reply:
splitted to LinkProcessors